### ISSUE TITLE
[Next 9.5.4+] Router.push duplicates query params in production mode, but not in dev mode

### ISSUE THREAD

**Issue by CarsonMcKinstry** [vercel/next.js#17840]
# Bug report

## Describe the bug

I recently came across this bug while using the "new" `router.push` method. Using the following

```
router.push({
    pathname: '/posts/[pid]',
    query: {
        pid: '1234',
        foo: 'bar'
    }
})
```

In dev, this directs me to `/posts/1234?foo=bar`

In prod, this directs me to `/posts/1234?pid=1234&foo=bar`

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. clone and install [this repo](https://github.com/CarsonMcKinstry/router-push-bug)
2. run `yarn build`
3. run `yarn start`
4. navigate to `localhost:3000`
5. press `The Button`

## Expected behavior

I would have expected production to run the same as in dev mode. Like I said above, `push`ing with this information

```
{
    pathname: '/posts/[pid]',
    query: {
        pid: '1234',
        foo: 'bar'
    }
}
```

Should send me to `/posts/1234?foo=bar`. 

## System information

- OS: macOS
- Browser (if applies): chrome
- Version of Next.js: 9.5.4, 9.5.5
- Version of Node.js: 12.16.1

**Comment by ShoeBoom** [comment]
I can work on this.

**Comment by ShoeBoom** [comment]
It seems that this bug exists as far back as version `9.0.0`. At this point I think the best solution is the document this behavior and fix the development version to comply instead.

Thoughts?

**Comment by erikdstock** [comment]
I spent about two hours faffing with this, particularly the [`resolveHref` method](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/next-server/lib/router/router.ts#L195-215) but ultimately couldn't get a consistent environment for debugging, logging, writing a test or making sense of the type coercion (like the selected lines above where we return a string or an array of strings). Probably started too deep trying to write a unit test for this function in retrospect. I was trying to find out why we only call these lines if `resolveAs` is passed:
```ts

    if (
          isDynamicRoute(finalUrl.pathname) &&
          finalUrl.searchParams &&
          resolveAs  // I think this is usually undefined when called from router.push()?
        ) {
      // throw new Error('HERE ' + finalUrl.search)
      const query = searchParamsToUrlQuery(finalUrl.searchParams)
      const { result, params } = interpolateAs(
        finalUrl.pathname,
        finalUrl.pathname,
        query
      )

      if (result) {
        interpolatedAs = formatWithValidation({
          pathname: result,
          hash: finalUrl.hash,
          query: omitParmsFromQuery(query, params),  // the only time we call omitParamsFromQuery()
        })
      }
    }
// ... execution continues until:
    return (resolveAs  // again, appears to be undefined
      ? [resolvedHref, interpolatedAs || resolvedHref]
      : resolvedHref) as string
```
Maybe a red herring but this was my suspicion - hard to debug with so many assumptions built in.

**Comment by ShoeBoom** [comment]
@erikdstock I think I know the source of the bug. [Here](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/next-server/lib/router/router.ts#L673)

It seems to occurs due to `resolvedAs `having different values in prod and dev

```js
    let resolvedAs = as   <-- "/posts/[pid]?pid=1234&foo=bar"

    if (process.env.__NEXT_HAS_REWRITES) {
      // this code is only run in dev
      // here dev will set resolvedAs  to "/posts/[pid]"
      resolvedAs = resolveRewrites(
        parseRelativeUrl(as).pathname, <------------------ "/posts/[pid]"
        pages,
        basePath,
        rewrites,
        query,
        (p: string) => this._resolveHref({ pathname: p }, pages).pathname!
      )

      if (resolvedAs !== as) {
        const potentialHref = removePathTrailingSlash(
          this._resolveHref(
            Object.assign({}, parsed, { pathname: resolvedAs }),
            pages,
            false
          ).pathname!
        )

        // if this directly matches a page we need to update the href to
        // allow the correct page chunk to be loaded
        if (pages.includes(potentialHref)) {
          route = potentialHref
          pathname = potentialHref
          parsed.pathname = pathname
          url = formatWithValidation(parsed)
        }
      }
    }
    // in dev resolvedAs = "/posts/[pid]" & and in prod resolvedAs = "/posts/[pid]?pid=1234&foo=bar"
    resolvedAs = delLocale(delBasePath(resolvedAs), this.locale)
```

This is because inside [resolveRewrites](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/next-server/lib/router/utils/resolve-rewrites.ts#L9)
```js
export default function resolveRewrites(
  asPath: string,
  pages: string[],
  basePath: string,
  rewrites: Rewrite[],
  query: ParsedUrlQuery,
  resolveHref: (path: string) => string
) {
  if (!pages.includes(asPath)) { <------------------- this is false
    // ....
  }
  return asPath <---------- // returns /posts/[pid]
}
```

**Comment by vitosamson** [comment]
I think I'm seeing this or a similar bug as well. I've created a minimal reproduction here: https://github.com/vitosamson/next-router-bug-repro. Spin it up (`yarn dev`), navigate to `/[id]` and then update the route via the Foo/Bar buttons.

For a route like `/[id]`, pushing a new route via `router.push({ pathname: '/[id]', query: { id: 'foo' } })` will cause `router.asPath` to become `/[id]?id=foo` instead of `/foo`.

This is causing issues for me when trying to update the route's search programatically (I'm not sure if there's another way I'm supposed to be doing that other than the `query` object?). I've had to resort to using `router.asPath` for the initial SSR and then `${window.location.pathname}${window.location.search}` thereafter when I need to know what the current full path is.

If this is fixed via #18009, any idea when a new version will be cut with that?
